VorpalEngine/include/vorpal/utility/asset/Storage.hpp

```diff
@@ -23,136 +24,141 @@
 
 namespace vp
 {
-    namespace utility
+namespace utility
+{
+namespace asset
+{
+/** @brief  Controls thread safe asset access
+ *
+ *  Provides both sync and async methods to request assets
+ *
+ *  @todo   cleaning up idle Handler object
+ */
+class Storage : public templates::Singleton<Storage>
+{
+   public:
+    typedef std::string KeyType;
+
+    /** @brief  Initializes worker threads
+     *
+     *  If upon the calling of this method there already were
+     *  some worker threads, they will be stopped and after that
+     *  given number of threads will be created
+     *
+     *  @param  count   number of worker threads
+     */
+    VORPAL_EXPORT void InitializeWorkers(uint16_t count);
+
+    /** @brief  Synchronously gets an asset identified by @p key
+     *
+     *  If asset is available, returns it. Otherwise:
+     *  - if the asset was not yet requested, prepares and returns it
+     *  - if the asset task is in the worker queue, tries to snatch it:
+     *  -- if task was snatched, prepares and returns the asset
+     *  -- if task was not snatched, waits until worker provides the asset
+     *
+     *  @param  key asset identifier
+     *
+     *  @return Handler shared object
+     */
+    VORPAL_EXPORT Handler Get(const KeyType& key);
+
+    /** @brief  Asynchronously gets an asset identified by @p key
+     *
+     *  If asset is avaiable, returns its shared future. Otherwise
+     *  - if the asset was not yet requested, prepares asset loading task
+     *      and pushes it to the worker queue
+     *  - if the asset task is in the worker queue, notifies it of @p priority;
+     *      (the biggest of priorities is taken)
+     *
+     *  @param  key         asset identifier
+     *  @param  priority    asset loading priority
+     *
+     *  @return shared future object on a Handler
+     */
+    VORPAL_EXPORT std::shared_future<Handler> GetAsync(
+        const KeyType& key, uint32_t priority = 100);
```

Comment:
Not a fan. Either all in one line or all in separate lines. Maybe wrapping column should be increased.